File: exercises/python/lab13/fork_sort.py
"""Lab 13 - Fork + parallel sort: parent uses insertion sort, child uses selection sort"""
import os
import sys
import time


def insertion_sort(arr):
    a = arr[:]
    for i in range(1, len(a)):
        key = a[i]
        j = i - 1
        while j >= 0 and a[j] > key:
            a[j+1] = a[j]
            j -= 1
        a[j+1] = key
    return a


def selection_sort(arr):
    a = arr[:]
    for i in range(len(a)):
        min_idx = i
        for j in range(i+1, len(a)):
            if a[j] < a[min_idx]:
                min_idx = j
        a[i], a[min_idx] = a[min_idx], a[i]
    return a


def main():
    print("=== Fork: Parallel Sorting ===\n")
    n = int(input("Enter number of elements: "))
    arr = list(map(int, input(f"Enter {n} elements: ").split()))

    pid = os.fork()

    if pid < 0:
        print("fork failed", file=sys.stderr)
        sys.exit(1)
    elif pid == 0:
        time.sleep(0.5)  # let parent print first
        sorted_arr = selection_sort(arr)
        print(f"\nChild  process (PID {os.getpid()}, PPID {os.getppid()}):")
        print(f"Selection Sort result: {sorted_arr}")
    else:
        sorted_arr = insertion_sort(arr)
        print(f"Parent process (PID {os.getpid()}):")
        print(f"Insertion Sort result: {sorted_arr}")
        os.wait()


if __name__ == "__main__":
    main()

File: exercises/cpp/lab13/fork_sort.cpp
// Lab 13 - Fork + parallel sort: parent uses insertion sort, child uses selection sort
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>

void insertionSort(int arr[], int n) {
    for (int i = 1; i < n; i++) {
        int temp = arr[i];
        int j = i - 1;
        while (j >= 0 && arr[j] > temp) {
            arr[j+1] = arr[j];
            j--;
        }
        arr[j+1] = temp;
    }
}

void selectionSort(int arr[], int n) {
    for (int i = 0; i < n - 1; i++) {
        int min = i;
        for (int j = i + 1; j < n; j++)
            if (arr[j] < arr[min]) min = j;
        int tmp = arr[i]; arr[i] = arr[min]; arr[min] = tmp;
    }
}

void printArray(int arr[], int n) {
    for (int i = 0; i < n; i++) printf("%d ", arr[i]);
    printf("\n");
}

int main() {
    int n;
    printf("=== Fork: Parallel Sorting ===\n\n");
    printf("Enter number of elements: ");
    scanf("%d", &n);

    int arr[30];
    printf("Enter %d elements: ", n);
    for (int i = 0; i < n; i++) scanf("%d", &arr[i]);

    // Copy for child
    int arr2[30];
    for (int i = 0; i < n; i++) arr2[i] = arr[i];

    pid_t pid = fork();

    if (pid < 0) {
        perror("fork");
        return 1;
    } else if (pid == 0) {
        // Child: selection sort
        sleep(1); // let parent print first
        printf("\nChild  process (PID %d, PPID %d):\n", getpid(), getppid());
        selectionSort(arr2, n);
        printf("Selection Sort result: ");
        printArray(arr2, n);
    } else {
        // Parent: insertion sort
        printf("Parent process (PID %d):\n", getpid());
        insertionSort(arr, n);
        printf("Insertion Sort result: ");
        printArray(arr, n);
        wait(NULL);
    }

    return 0;
}
